### PR TITLE
fix(scope-manager): correct analysis of inferred types in conditional types

### LINKED ISSUE #2526: [@typescript-eslint/no-shadow] false positives with conditional types and `infer`
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-shadow": ["error"]
  }
}
```

```TS
export type ArrayInput<Func> = Func extends (arg0: Array<infer T>) => any
  ? T[] // if true
  : Func extends (...args: infer T) => any // else
    ? T
    : never;

function f1(...s: string[]) {
  return s.length;
}

function f2(numbers: number[]) {
  return numbers.length;
}

export let r1: ArrayInput<typeof f1>;
export let r2: ArrayInput<typeof f2>;
```

**Expected Result**

No shadowing should be reported.  The first `infer T` is only in scope in the first "if true" expression.  In the "else" expression, it is valid to use `infer T` again, as the earlier `T` is not in scope.

**Actual Result**

A false positive is reported:

> "'T' is already declared in the upper scope."

**Versions**

This was reproduced in both 4.0.2 and 4.1.0.

### PR DESCRIPTION
**Pull request by dinofx**
This fixes #2526 
In a conditional type, the inferred types are not in scope in the "if false" type.

### CODE REVIEW

File: packages/scope-manager/src/referencer/TypeVisitor.ts [hunk 1]
```diff
@@ -94,9 +94,11 @@ class TypeVisitor extends Visitor {
     // which are only accessible from inside the conditional parameter
     this.#referencer.scopeManager.nestConditionalTypeScope(node);
 
-    this.visitChildren(node);
+    this.visitChildren(node, ['falseType']);
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/TypeVisitor.ts:97]:
```suggestion
    // type parameters inferred in the condition clause are not accessible within the false branch
    this.visitChildren(node, ['falseType']);
```

**Review by bradzacher — CHANGES_REQUESTED** [review]
Fix itself LGTM!

----

To help prevent regressions, could you please add a new snapshot test into `packages/scope-manager/tests/fixtures/type-declaration`
Call it something like `conditional-nested.ts`

The test should be something like:

```ts
type Test<T> = T extends Array<infer U>
  ? U
  : T extends Set<infer U>
  ? U
```

Then you can `cd pacakges/scope-manager && yarn test fixtures -u` to update the snapshot.
What we're looking for here is a snapshot wherein there are two separate `U` variables created each with their own reference.

If you're not sure what to look for, just push the snapshot and I will be able to verify it.

**Comment by dinofx** [comment]
Added the new fixture and snapshot

**Comment by dinofx** [comment]
@bradzacher are any other changes needed?

**Review by bradzacher — APPROVED** [review]
LGTM - thanks for doing this!

**Pull request merged**